src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -449,28 +450,40 @@ def _compile_hermetic(self, jvm_options, ctx, classes_dir, jar_file, zinc_args,
       additional_snapshots +
       (self.extra_resources_digest(ctx),)
     )
-
-
-    argv = image_specific_argv + zinc_args
-    # TODO(#6071): Our ExecuteProcessRequest expects a specific string type for arguments,
-    # which py2 doesn't default to. This can be removed when we drop python 2.
-    argv = [text_type(arg) for arg in argv]
-
-    req = ExecuteProcessRequest(
-      argv=tuple(argv),
-      input_files=merged_input_digest,
-      output_files=(jar_file,) if self.get_options().use_classpath_jars else (),
-      output_directories=() if self.get_options().use_classpath_jars else (classes_dir,),
-      description="zinc compile for {}".format(ctx.target.address.spec),
-      jdk_home=self._zinc.underlying_dist.home,
-    )
-    res = self.context.execute_process_synchronously_or_raise(
-      req, self.name(), [WorkUnitLabel.COMPILER])
-
-    # TODO: Materialize as a batch in do_compile or somewhere
-    self.context._scheduler.materialize_directories((
-      DirectoryToMaterialize(get_buildroot(), res.output_directory_digest),
-    ))
+    BATCH_SIZE = 10
+    def batch(list_to_batch, n=1):
+      l = len(list_to_batch)
+      for chunk in range(0, l, n):
+        yield list_to_batch[chunk:min(chunk + n, l)]
+
+    original_zinc_args = zinc_args
+
+    for batched_sources in batch(sources, BATCH_SIZE):
```

Comment:
The batching looks good, but in order for all of the `ExecuteProcessRequest`s to run in parallel, you'll have to collect all of them into a list, and then invoke them via one call to the scheduler. That might look like:

```
reqs = []
for batched_sources in ...:
  ..
  reqs.append(ExecuteProcessRequest(..))
...
results = self.context._scheduler.product_request(ExecuteProcessResult, reqs)
```

And then you'll need to take all of the `ExecuteProcessResult`s you get back, and "merge them" via `self.context._scheduler.merge_directories`. Merging is easy if `!self.get_options().use_classpath_jars`.. it's a little harder if a jar is in use... for a first iteration of this patch, I would suggest just blowing up if `use_classpath_jars` is true, and we can discuss how to fix it before landing.